rfcs/0000-central-station.md

```diff
@@ -0,0 +1,507 @@
+handling logic is about deciding what trip a vehicle is on.
+
+It will also consume the ServiceChanges feed, which is how CentralStation will convey the TSCH
+offsets it receives.
+
+As a bonus, by relying on GTFS trips and ServiceChanges, RTR won't have to rely on handcrafted
+patterns, and should be able to predict, e.g., new stations or closed stations automatically.
+
+Question: how do you make reverse predictions this way? Is there something in this data model that
+allows us to convey to RTR the _subsequent_ trip for a vehicle?
+
+## Busloc
+
+### Current description
+
+Busloc primarily keeps track of the whereabouts of MBTA buses and publishes a few feeds related to
+them. Busloc runs on the MBTA network because one of the sources of its data, the TransitMaster
+DB, is only accessible from there.
+
+The majority of what Busloc deals with is "normal" bus data. These are buses logged on and running
+their expected, planned routes. Busloc gets this data from TransitMaster, via its hacky internal
+spelunking of the TransitMaster DB, and via Samsara, a cloud service that has more frequent GPS
+updates. Busloc combines these feeds, taking the latest lat/long from either source for every
+vehicle. These logged-on vehicles have their GTFS trip directly in the data, since TransitMaster
+and GTFS both get their source data from HASTUS. The Samsara data doesn't _have_ to be queried
+from the MBTA network, but is done to have all the data in one place and for simplicity merging
+with the TransitMaster data. This data is published as VehiclePositions, which is consumed by
+Swiftly to generate TripUpdates. The data is GTFS-rt-esque, but includes block and run data, which
+Swiftly takes advantage of.
+
+In addition to the planned service, TransitMaster also has "overload" trips, which is when a bus
+is assigned to run a route that it wasn't planned for. These create "ADDED" trips, based on a
+scheduled trip but with an offset at every location. Since Swiftly cannot produce predictions for
+non-GTFS trips (is this still true?), Busloc marks these with a special "unassigned" status,
+preventing Swiftly from incorrectly matching them to a GTFS trip. TransitMaster also generates
+predictions (of unknown quality, or perhaps simply the arrival/departure values from the
+underlying trip + offset), and these are published in a TripUpdates file which concentrate
+consumes directly.
+
+Conversely, "block waivers", when trips are canceled, are also handled by Busloc and also bypass
+Swiftly. Busloc receives the list of waivers via TransitMaster and publishes them as a special
+TripUpdates file, marking their stops as "SKIPPED". Swiftly is meanwhile publishing predictions
+for the canceled trips because the VehiclePositions are in the feed, and Swiftly can't handle
+waivers. So concentrate merges the Swiftly TripUpdates and the special Busloc TripUpdates feed and
+cancels the trips there. (Is that correct?)
```

Comment:
Note that this information is also consumed by Skate to generate notifications (we take StopTimeUpdates and roll them up into individual notifications as appropriate).